gateway/src/apicast/gc.lua

```diff
@@ -0,0 +1,53 @@
+-- In LuaJIT and Lua 5.1, the __gc metamethod does not work in tables, it only
+-- works in "userdata". This module introduces a workaround to make it work
+-- with tables.
+
+local getmetatable = getmetatable
+local newproxy = newproxy
+local ipairs = ipairs
+local pairs = pairs
+local tostring = tostring
+
+local _M = {}
+
+local function __gc(proxy)
+  local t = getmetatable(proxy).__table
+  local mt = getmetatable(t)
+
+  if mt and mt.__gc then mt.__gc(t) end
+end
+
+local function __call(proxy, ...)
+  local t = getmetatable(proxy).__table
+  local mt = getmetatable(t)
+
+  if mt and mt.__call then
+    return mt.__call(...)
+  end
+end
+
+function _M.enable(table)
+  -- newproxy() returns a userdata instance
+  local proxy = newproxy(true)
+
+  -- We are going to define a metatable in the userdata instance to make it act
+  -- like a table. To do that, we'll just define the metamethods a table should
+  -- respond to.
+  local mt = getmetatable(proxy)
+
+  mt.__gc = __gc
+  mt.__call = __call
+
+  mt.__index = table
+  mt.__newindex = table
+  mt.__table = table
```

Comment:
I think it would be nice to also hide the this metatable by:

```lua
mt.__metatable = getmetatable(table)
```

Then `getmetatable(proxy)` will return the metatable returned by `getmetatable(table)`. 

This would not propagate if someone would change the metatable of `table` directly, but that can be avoided by changing the API a bit.